fe/fe-core/src/main/java/com/starrocks/sql/optimizer/InputPropertyDeriver.java

```diff
@@ -0,0 +1,290 @@
+        this.requirements = requirements;
+
+        inputProperties = Lists.newArrayList();
+        groupExpression.getOp().accept(this, new ExpressionContext(groupExpression));
+        return inputProperties;
+    }
+
+    @Override
+    public Void visitOperator(Operator node, ExpressionContext context) {
+        List<PhysicalPropertySet> inputProps = new ArrayList<>();
+        for (int childIndex = 0; childIndex < context.arity(); ++childIndex) {
+            inputProps.add(PhysicalPropertySet.EMPTY);
+        }
+        inputProperties.add(inputProps);
+        return null;
+    }
+
+    @Override
+    public Void visitPhysicalHashJoin(PhysicalHashJoinOperator node, ExpressionContext context) {
+        String hint = node.getJoinHint();
+
+        // 1 For broadcast join
+        PhysicalPropertySet rightBroadcastProperty =
+                new PhysicalPropertySet(new DistributionProperty(DistributionSpec.createReplicatedDistributionSpec()));
+        inputProperties.add(Lists.newArrayList(PhysicalPropertySet.EMPTY, rightBroadcastProperty));
+
+        ColumnRefSet leftChildColumns = context.getChildOutputColumns(0);
+        ColumnRefSet rightChildColumns = context.getChildOutputColumns(1);
+        List<BinaryPredicateOperator> equalOnPredicate =
+                getEqConj(leftChildColumns, rightChildColumns, Utils.extractConjuncts(node.getOnPredicate()));
+
+        if (Utils.canOnlyDoBroadcast(node, equalOnPredicate, hint)) {
+            return null;
+        }
+
+        if (node.getJoinType().isRightJoin() || node.getJoinType().isFullOuterJoin()
+                || "SHUFFLE".equalsIgnoreCase(hint) || "BUCKET".equalsIgnoreCase(hint)) {
+            inputProperties.clear();
+        }
+
+        // 2 For shuffle join
+        List<Integer> leftOnPredicateColumns = new ArrayList<>();
+        List<Integer> rightOnPredicateColumns = new ArrayList<>();
+        JoinPredicateUtils.getJoinOnPredicatesColumns(equalOnPredicate, leftChildColumns, rightChildColumns,
+                leftOnPredicateColumns, rightOnPredicateColumns);
+        Preconditions.checkState(leftOnPredicateColumns.size() == rightOnPredicateColumns.size());
+
+        HashDistributionSpec leftDistribution = DistributionSpec.createHashDistributionSpec(
+                new HashDistributionDesc(leftOnPredicateColumns, HashDistributionDesc.SourceType.SHUFFLE_JOIN));
+        HashDistributionSpec rightDistribution = DistributionSpec.createHashDistributionSpec(
+                new HashDistributionDesc(rightOnPredicateColumns, HashDistributionDesc.SourceType.SHUFFLE_JOIN));
+
+        PhysicalPropertySet leftInputProperty = createPropertySetByDistribution(leftDistribution);
+        PhysicalPropertySet rightInputProperty = createPropertySetByDistribution(rightDistribution);
+
+        inputProperties.add(Lists.newArrayList(leftInputProperty, rightInputProperty));
+        return null;
+    }
+
+    private Optional<HashDistributionDesc> getRequiredLocalDesc() {
```

Comment:
Why need the method in InputPropertyDeriver？

Reply:
removed